### ISSUE TITLE
cli/extract: update source translations

### ISSUE THREAD

**Issue by snegostup** [lingui/js-lingui#199]
This is obviously only relevant when using messages with custom IDs.

Right now if we `extract` messages and then decide to update them (in source code) and run `extract` again - it won't update the messages in catalog. It will leave initial strings and there'd be no way to update them other than manually edit the catalog.

I think it shouldn't be like that and we should update source catalog with updated messages in this case.

We can change [this string](https://github.com/lingui/js-lingui/blob/fa7adde33988a85c7d1faed1aba50e80b9a398e4/packages/cli/src/api/formats/lingui.js#L73):
```javascript
translation: prevCatalog[key].translation,
```
to
```javascript
translation: config.sourceLocale === locale ? nextCatalog[key].defaults : prevCatalog[key].translation,
```
**And this leads to another issue** :) (I wrote about it [here](https://github.com/lingui/js-lingui/issues/141#issuecomment-374518007)) - when using `minimal` format with custom IDs, it extracts as `'message.id': 'message.id'` instead of `'message.id': 'default message'`
This can be fixed [here](https://github.com/lingui/js-lingui/blob/34b866940e9de832b3b52928361d456818470746/packages/cli/src/api/formats/lingui.js#L64), changing: 
```javascript
translation: config.sourceLocale === locale ? key : "",
```
to
```javascript
translation: config.sourceLocale === locale ? message.defaults : "",
```
What do you think?

**Comment by tricoder42** [comment]
Sounds reasonable. I'm gonna take a look at this issue today. There seems to be many related issues.

**Comment by tricoder42** [comment]
@snegostup I've just created a PR, would you be interested in review? I updated your proposed solution little bit:

Instead of

```js
translation: config.sourceLocale === locale 
  ? nextCatalog[key].defaults 
  : prevCatalog[key].translation,
```

there is

```js
translation: config.sourceLocale === locale && 
prevCatalog[key].translation === prevCatalog[key].defaults 
  ? nextCatalog[key].defaults 
  : prevCatalog[key].translation,
```

to avoid loosing custom translations (even in catalog for `sourceLocale`).

**Comment by snegostup** [comment]
Thanks!

> when using custom message IDs and minimal format, once catalog for sourceLocale is generated translations must be always edited manually. Even when default translation in source code is updated, the translation in catalog won't be updated automatically because there's no way to distinguish default translations from custom ones.

This is really a problem. Like a coder I'm not comfortable having multiple sources of truth (code and catalog) and would gladly have an `--overwrite` option that would overwrite all string in favor of ones from code. On the other hand I can imagine the case when proofreader corrects my ugly English strings (in catalog) and I need to keep them. I need to think a bit more.

**Comment by tricoder42** [comment]
It is indeed suboptimal.

**What if** we add interactive mode to lingui extract? When different `defaults` are encountered it either means the source code has changed or proofreader changed the catalog. CLI should stop collecting messages and give you two options:
- either update message in source code (manually)
- or update `defaults` in catalog

Options would be: `code`, `catalog`, `all from code`, `all from catalog` and the last two could probably have CLI options (`--update-defaults`, `--keep-defaults`)

**Comment by tricoder42** [comment]
In simple cases, we could also update message in code automatically. Complex messages like `{value, plural, one {Book} other {<strong>#</strong> Books}}` would have to be updated manually anyway (auto update seems to me too complex at the moment 😄 )

**Comment by snegostup** [comment]
Great idea. And I agree that we don't need to update code automatically, just display good message and exit. It reminds me how DBMS would react if you try to create unique index on some field and it encounters two records with same value of the field. It would just display error message and stop. Most probably you already know why the messages are different and can quickly update while autoupdate would certainly be error prone and would need a lot of polishing.

This also reminds me of #200 -- we can also throw similar error message and stop when encountered different messages with same ID within code

**Comment by snegostup** [comment]
:+1:

**Comment by tricoder42** [comment]
Fixed and released in [v2.2.0](https://github.com/lingui/js-lingui/blob/stable-2.x/CHANGELOG.md#220-2018-07-04). Thanks to all involved!
